### ISSUE TITLE
Trouble training agent

### ISSUE THREAD

**Issue by AidanRocke** [pauli-space/tic-tac-toe_RL#1]
After the first iteration of training, that is just one game, the policy network starts outputting nans. Not sure why this is the case right now but my hunch is that the problem might be here:

```python
def multinomial(self):
        
        ## identify the free positions:
        free_positions = tf.to_float(tf.equal(self.state,tf.zeros((1,9))))

        free_matrix = tf.diag(tf.reshape(free_positions,(9,)))
    
        ## calculate probability vector:
        prob_vec = tf.transpose(tf.matmul(free_matrix,tf.transpose(self.policy)))
        prob = prob_vec/tf.reduce_sum(prob_vec)
        
        return tf.contrib.distributions.Multinomial(total_count=1., probs=prob)
```

Actually I managed to duplicate this error with a variation of the above script. This appears to be the show-stopping error:

> InvalidArgumentError (see above for traceback): Input to reshape is a tensor with 81 values, but the requested shape has 9
> 	 [[Node: Reshape_1 = Reshape[T=DT_FLOAT, Tshape=DT_INT32, _device="/job:localhost/replica:0/task:0/device:CPU:0"](ToFloat_1, gradients/Add_grad/Shape_1)]]

**Comment by AidanRocke** [comment]
The solution to this problem required using the map function as follows:

```
def multinomial(self):
        
        ## identify the free positions:    
        free_positions = tf.to_float(tf.equal(self.state,tf.zeros((1,9))))
    
        fm_mapping = lambda x: tf.diag(tf.reshape(x,(9,)))
    
        free_matrices = tf.map_fn(fm_mapping,free_positions)


        ## calculate probability vector:
        pvec_mapping = lambda x: tf.transpose(tf.matmul(x,tf.transpose(self.policy)))
        
        prob_vec = tf.map_fn(pvec_mapping,free_matrices)
        prob = prob_vec/(tf.reduce_sum(prob_vec)+tf.constant(1e-5))

    
        return tf.contrib.distributions.Multinomial(total_count=1., probs=prob)
```
